Improve API and Web interface

DRSDavidSoft opened the pull request:
I'm working to improve the MPC-HC's API and Web interface according to the changes in my personal fork, and port my changes changes in order to upstream them.

I would appreciate it if you could have a review. Sorry for having multiple different aspects all integrated in a single PR.

### Added features
- [x] Added _new_ API to query (and set) the current volume: `CMD_CURRENTVOLUME` and `CMD_GETVOLUME`  
    (This will be automatically sent upon changes and can also be queried from the host side)
    To set the volume, the `CMD_SETVOLUME` with syntax: `<volume>|<is_muted>` can be used.
- [x] Added dark mode to the Web interface and some improvements  
    (e.g. the `variables.html` page now also displays the variable name in addition to its value using CSS)

### Fixes and general changes
- [x] Replaced the repeated volume get logic with pre-existing `GetVolume()` and `IsMuted()` to make code DRY
- [x] Added simple process start error handling to MPCTestAPI app
- [x] Updated MPCTestAPI app to handle paths with spaces
- [x] Updated the MPCTestAPI app to handle the Enter key correctly
- [x] Updated MPCTestAPI to add some missing API items, as well as better handling the unknown codes:
    - CMD_DISCONNECT
    - CMD_NOTIFYSEEK
    - CMD_CURRENTVOLUME (new)
    - CMD_VERSION
    - ...
- [x] Updated MPCTestAPI's version and copyright year (since we're making changes to it)
- [x] Updated some comments in the `MpcApi.h` for clarity

### Changes to be made
- [ ] Prevent firing `OnPlayVolume()` with duplicated values  
    To replicate this issue, open the MPCTestAPI app and invoke increase/decrease volume commands.
    This will cause the event to be fired twice, which is not desirable.
    Additionally, dragging the volume bar's handle past zero will fire with the same `0` value repeatedly.

- [ ] Add a new API to query the currently connected host hWnd
    This is useful for scripting purposes and/or when dealing with multiple instances (windows) of MPC-HC
    Allowing us to query whether a particular MPC-HC window has a host connected or not.

- [ ] Add a new API to display status messages using the API

- [ ] ~Add the missing functionality of sending OSD and Status messages to the MPCTestAPI app~  
    (Since the parameters format differ from other API commands, this will be skipped)

### Edit
Will move some unrelated feature/improvement tasks to their own issue later:

- [ ] Make use of Windows 11 rounded corners for the custom dark mode context menus (fix #2141)  
    (Currently they are not rounded)

- [ ] Re-write the web interface (and the HTTP server) to make use of Web-sockets instead of polling interval to improve user-experience
    Completely drop support for Microsoft IE  
    (Unless someone is still using it instead of modern web browsers)

DRSDavidSoft committed:
Ignore /src/MPCTestAPI/.vs

DRSDavidSoft committed:
Safe path handling
When a certain specified path contains spaces (such as "..\..\..\..\bin\mpc-hc_x64_Debug Lite\mpc-hc64.exe" that contains a space between "Debug" and "Lite"), then double quotes are required to safely quote the path

DRSDavidSoft committed:
Add simple error handling on process start

DRSDavidSoft committed:
Display unknown command code; Add missing CMD_DISCONNECT case

Updated `GetMPCCommandName` to include a case for `CMD_DISCONNECT`. Enhanced the default case to return a formatted string with the command's hexadecimal value for better debugging of unknown commands, replacing the previous static "CMD_UNK" return value.

DRSDavidSoft committed:
Add missing CMD_NOTIFYSEEK case to GetMPCCommandName

DRSDavidSoft committed:
Update version to 1.0.2.0 and copyright year to 2025

Updated the version number of the "MPC-HC Test API" from 1.0.1.0 to 1.0.2.0 in both the dialog resource and the version information block. Changed the copyright year from 2013 to 2025 in the dialog resource and the version information block.

DRSDavidSoft committed:
Add CMD_VERSION handling and version retrieval command

Updated `GetMPCCommandName` to include `CMD_VERSION` case
and modified `OnCopyData` to automatically send `CMD_GETVERSION`
upon receiving `CMD_CONNECT`.

DRSDavidSoft committed:
Add CMD_NOTIFYENDOFSTREAM to GetMPCCommandName

DRSDavidSoft committed:
Fix indentation for variable in HTML

Aligned the `<p id="audiotrack">[audiotrack]</p>` element
by removing spurious leading space to match the indentation
of other elements in the file.

DRSDavidSoft committed:
Improve comment

DRSDavidSoft committed:
Improve comments

DRSDavidSoft committed:
Add API enum for retrieving volume; cleaned-up comments
- Added new options to request and receive the current MPC-HC volume through the API.
- Cleaned-up some comments.

DRSDavidSoft committed:
Update version information

DRSDavidSoft committed:
Add API to send current volume
- Updated the test application with new combobox items
- The current volume can be requested by `CMD_GETVOLUME` or sent automatically

DRSDavidSoft committed:
Use `GetVolume()` and `IsMuted()` instead of control's implementation detail
These methods existed in the class yet in these places the code was still using the repeated code.
The existing methods return the same value.

DRSDavidSoft committed:
Improved web interface
- Dark mode support
- Default viewport for mobile
- Transparency for rounded title corners
- Added note regarding future updates to index

DRSDavidSoft committed:
Refactor process handling and add OnOK function

- Fixed the order of `UpdateData(TRUE);`, it must come before accessing `m_strMPCPath` otherwise the data is not valid yet and you needed to call the function twice for it to take effect.
- Updated the OnButtonFindwindow function to format the command line string
- Added a simple OnOK function to handle Enter key based on the focused control

clsid2 commented on src/MPCTestAPI/MPCTestAPIDlg.cpp:
```diff
@@ -44,8 +46,21 @@ LPCTSTR GetMPCCommandName(MPCAPI_COMMAND nCmd)
+            static CString strResult;
+            LPCTSTR pszName = nullptr;
+            pszName = _T("UNKNOWN");
+            strResult.Format(_T("%s (0x%08X)"), pszName, (unsigned int)nCmd);
```
Use UNKNOWN directly in the Format string.

DRSDavidSoft replied:
Fixed by https://github.com/clsid2/mpc-hc/pull/3403/commits/eda3efc5ca38750d1fb8bc5c3ac835ca7f9fd629

clsid2 commented on src/mpc-hc/MainFrm.cpp:
```diff
@@ -20531,6 +20536,20 @@ void CMainFrame::SendCurrentPositionToApi(bool fNotifySeek)
+        return;
+    }
+
+    CString buff, volumelevel, muted;
```
A single string can be used here.

DRSDavidSoft replied:
Fixed by: https://github.com/clsid2/mpc-hc/pull/3403/commits/808aa2bc5fa73ba08b184ba24124ad5ee082c847

clsid2 commented on src/mpc-hc/res/web/img/headerclose.png:
Why did these two png files suddenly get much bigger?

DRSDavidSoft replied:
Good question; the original files were optimized (if you see the git history) and I neglected to do so when I added transparent corners to the new image files. This has been fixed with https://github.com/clsid2/mpc-hc/pull/3403/commits/cc33aeef00e4b95ba2b8da8ad8d47ab87828b77b

clsid2 commented:
Future changes unrelated to MPCTestAPI should be in separate PRs.

DRSDavidSoft committed:
fix: Use UNKNOWN directly in the Format string.

clsid2 commented on src/MPCTestAPI/MPCTestAPIDlg.cpp:
`strResult.Format(_T("UNKNOWN (0x%08X)"), (unsigned int)nCmd);`

DRSDavidSoft committed:
fix: A single string can be used here.

DRSDavidSoft committed:
Optimized png filesize
Removed XMP metadata to reduce filesize as well

DRSDavidSoft commented:
> Future changes unrelated to MPCTestAPI should be in separate PRs.

Sure, in this PR I was mainly focused on adding the MPC-HC API for querying the volume value. I also added it to the MPCTestAPI app.

Since there were some bugs in the original MPCTestAPI, I thought to just commit them here as well as some minor tweaks.

In the future, I'll open a separate PR for any MPCTestAPI related commits that are not directly related to the main MPC-HC app.

I'll now focus on the remaining tasks in this PR.

DRSDavidSoft commented:
> * Prevent firing `OnPlayVolume()` with duplicated values
>   To replicate this issue, open the MPCTestAPI app and invoke increase/decrease volume commands.
>   This will cause the event to be fired twice, which is not desirable.
>   Additionally, dragging the volume bar's handle past zero will fire with the same `0` value repeatedly.

I'm going to just use a `m_lastVolume` variable to keep track of the last reported volume value, and only report the new value if it was changed, if this is OK with you @clsid2

DRSDavidSoft committed:
Introduce new `CMD_SETVOLUME` API for volume and mute control

- Adds a new command `CMD_SETVOLUME` to manage the current volume and mute state
- The command is defined in `MpcApi.h` and implemented in `CMainFrame::ProcessAPICommand` to handle volume and mute updates
- The `CPlayerToolBar` class has been updated to provide access to methods for mute state management
- Additionally, a conflict with the how `CMD_SETVOLUME` was referenced in the web API has been fixed, where it has been renamed to `CMD_SETVOL` for consistency (It's recommended to rename such cases to `WEBCMD_...` instead to prevent such collisions)
- Resource file `MPCTestAPI.rc` has also been updated to add new combobox items in the MPCTestAPI app

DRSDavidSoft commented:
Added a new commit (https://github.com/clsid2/mpc-hc/pull/3403/commits/6544c0294ae2d80ea05e3552424ffd34f0af47d7) that allows changing the volume from the API, which was previously missing.

Note: This is kind of off-topic; but sadly there is a noticeable delay from when the volume is changed until is actually reflected, doesn't matter the source (if it was set from either Web or regular API, or from the UI). I hope to investigate and see if I can make this instantaneous for my personal use case.

clsid2 commented:
I don't like the combination of volume and mute in a single parameter. Just keep it simple and set only volume. For practical purposes a volume of 0 has the same effect a muting.

DRSDavidSoft commented:
> I don't like the combination of volume and mute in a single parameter.

I assumed having the same syntax for both get/set API (passing multiple parameters for volume/info) is more in line with how the other existing APIs are implemented.

> Just keep it simple and set only volume. For practical purposes a volume of 0 has the same effect a muting.

I tried it out before making the commit; but @clsid2, I believe it's not the case, setting the volume to 0 is not actually the same as having the mute button turned on.

It's also a minor problem in the way the Web API reports the volume, which is not actually correct.

The API should reflect the actual state of the form, where the **volume level** and **mute state** are two separate things and you can modify one without changing the state of the other.

For example, I can set the volume to 50%. Then I can toggle the mute state. When the mute button is toggled, the volume slider won't jump to 0, it still shows 50%.

This behavior is desirable in the UI, and is excellent.

Now, when we are interfacing with the API to make changes, we should be able to toggle the mute state, without actually changing the volume level, or vice versa, change the volume level without affecting the mute toggle.

Hence, instead of making a separate API to modify the mute state, I made use of the second parameter for this use case.

And as I described above, using other values like 0 for mute won't work because it's not actually turning on the mute button, it's just setting the volume slider to 0.

Using a value of -1 to turn on mute is indeed consistent with the internal coding, so which one do you prefer?

Note that using APIs it should be intended behavior to be different than to how you recently implemented CLI.

Using CLI it's intended behavior to automatically turn off mute in case a volume switch was passed by user.

Using the API, it's up to the host program to control each aspect of the volume/mute values.

### Option A (current)
Syntax: `<volume>|<mute>`

Examples:
- `55|0` sets the volume slider to 55% and disables mute
- `90|1` sets the volume slider to 99% and enables mute 
- `30` sets the volume to 30%, doesn't affect the mute state:
    - If it's already muted, then it will stay muted but the volume slider will be set to 30 for when the mute button is tuend off
    - if it's not already muted then the volume will be updated to 30
- `|1` Passing this will enable mute, whatever the current volume is
- `|0` Passing this will turn off the mute, without changing whatever the current volume is

Note that in the last (`|0` and `|1`) cases the first parameter is skipped, and since it wasn't provided, then MPC-HC won't adjust the volume slider.

This is desirable behavior because optionally you'd like to have independent control over volume and mute values.

I also don't 100% enjoy that the syntax for controlling the mute state independently is like this, but it was the simplest and most consistent to implement, and as I said it's up to the host program to make use of this.

### Option B
Now, as you suggested, we can forget about having two parameters for the SET command and only use a single parameter, like this:

- `<volume>`: Set the volume and always disable mute
- `-1`: Turns the mute state on

This is more close to how you recently implemented the volume switch. I'd also argue that it's a good syntax but has the caveat of being inconsistent with the GET syntax.

To put an emphasis once again:
- Passing **0** will _NOT_ enable the mute button, it'll just set the volume slider to 0, which sure, _seems_ like it's muted, but it's actually not.
- Passing **-1** will actually turn on the mute button without modifying the volume slider, which is desirable.

This is once again, a glaring problem in the way the Web API currently reports the volume, and I plan to fix it as well in the new, modem websocket-based web API that I'm going to open a PR for.

So, which option would you prefer to have?

In my opinion both A and B are good, you trade consistency with removing the second parameter.

And I'd like to once again emphasize that in the GET API, we don't want to loose the information of the actual current volume and the independent mute state.

Hopefully I can also update the other places the volume is reported (like OSD and the Logitech keyboard LCD), if they're needed.

clsid2 commented:
option C, separate command for mute.

Mute state is already correctly reflected in web interface and its variables.

DRSDavidSoft commented:
OK. That may require four sets of commands, is this the correct approach? To be exact, which one are you suggesting?

### Option C-1:
- a single GET that reports _both_ volume and mute state back (as it currently is)
- a single SET for just volume
- a single SET for just mute

### Option C-2:
- a single GET for volume
- a single GET for mute state
- a single SET for volume
- a single SET for mute state

Note that option C-2 will require sending both back to the host. I feel like adding so many commands is starting to bloat the API for something as simple as volume/mute management...

---

Personally, I'd still like to ask you and suggest going with option A (since in the implementation, volume and mute state are linked). But if option C is the only one that you'll accept, I'll implement that one instead.

clsid2 commented:
C2

DRSDavidSoft commented:
OK, I'll make a new commit to make the mute state separate from volume for both GET and SET.

**Edit:** Investigating this, I see some issues that needs to be addressed.

DRSDavidSoft committed:
Merge branch 'clsid2:develop' into develop

DRSDavidSoft commented:
@clsid2, I'm not quite sure about closing this draft as fully superseded by #4053 and ask that #3403 be reopened for review.

#4053 did useful work and I have retained it rather than replaying its web changes. However, its own merged commit says that the API changes from #3403 were not included. A current-tree comparison confirms that it did not include the native `WM_COPYDATA` volume/mute API or any of the MPCTestAPI fixes. The closure event itself contains no stated reason or associated commit, so “fully superseded” is technically too broad: #4053 superseded the web subset, not the PR's remaining API/test scope.

I have now rebuilt the branch from current upstream `develop` (there is no `main` branch), backed up the previous tip, and completed the requested C2 design:

- separate GET volume, GET mute, SET volume, and SET mute commands;
- setting one preserves the other;
- volume 0 and the mute toggle are represented independently;
- explicit GET always replies, while automatic notifications are deduplicated;
- all toolbar, classic/themed slider, web, and API setter paths now converge on one real-change notification;
- new high-frequency callbacks use a bounded timeout;
- #4053's existing web endpoint still uses the same `-2` pseudo-command, with only an internal rename to avoid the native enum collision.

This also fixes a concrete issue left in #4053: `IsMuted()` previously inferred mute from the attenuated volume, so an unmuted slider at 0 was exposed as muted by `/status.json`, the remote, and classic variables. It now reads the actual toggle state.

The original MPCTestAPI fixes are restored and hardened: quoted/current paths, actionable process errors, safe Enter routing, validated/bounded `WM_COPYDATA`, version query on connect, a data-driven command map, exhaustive response labels, and the exact requested unknown format. The unfinished host-query and status-message checklist entries are implemented too. WebSocket and rounded-menu work remain separate, per the earlier scope request.

Validation on this machine:

- `git diff --check`: clean
- MPCTestAPI Release x86/x64: passed
- MPC-HC Release x86/x64 project compile/link: passed
- focused runtime harness: 55/55 checks passed, including the C2 state matrix, forced replies, deduplication, live/no-host host queries, and status display/expiry
- the whole solution still reports an external LAV toolchain failure (matching x86 MinGW / gnutls is unavailable locally), not an error in these projects

I have updated the PR checklist to distinguish implemented work, work already superseded by #4053, and work deliberately moved out. Please reopen #3403 so the remaining native API and MPCTestAPI work can be reviewed. If you prefer the host-query/status additions split from the C2 volume/mute recovery, I can split them into a separate commit or PR; please state the desired boundary rather than treating the excluded API work as already merged.

@adipose, thank you for the useful web fixes and remote work in #4053. I have preserved those changes and avoided reintroducing the old web pages. I am asking only that its supersession claim be narrowed to what it actually carried.

The inspectable recovery is the [three-commit compare against current `develop`](https://github.com/clsid2/mpc-hc/compare/develop...DRSDavidSoft:mpc-hc:develop), ending at [`75f948b20`](https://github.com/DRSDavidSoft/mpc-hc/commit/75f948b201a43bb59123107bdc0a7ccbea450244). GitHub currently keeps the closed PR's displayed head SHA pinned to the pre-recovery tip even though the fork branch has advanced; reopening should refresh it. If GitHub does not reattach the new head after reopening, I will open the same three commits as a new focused PR instead.

DRSDavidSoft commented:
@clsid2, @adipose: I have completed the requested domain split; can you please re-open this PR?

- #4062 — MPCTestAPI hardening
- #4065 — C2 volume/mute API and non-atomic client-composed state example
- #4064 — transient status messages
- #4061 — connected-host query
- #4063 — transparent legacy header assets

I updated this PR's body/checklists to map the complete recovery into those PRs and #4053, including the landing/rebase order and validation boundary. Every focused branch starts at the same current upstream `develop` SHA, passed a clean-scope/diff audit, and is authored from the preserved work rather than replaying #4053's web changes.

@adipose, thank you for confirming on #4053 that the API changes had not been ported and for offering to review them. I reacted there and replied with the final links. The useful #4053 web work remains intact; #4065 supplies the excluded native API and corrects the volume-zero/actual-mute distinction feeding its status JSON/remote.

For clarity, my technical objection to the original closure rationale still stands: #4053 did not fully supersede the API/test scope. But now that each domain has its own open PR, #3403 can remain closed as the historical integration/reference record rather than being reopened for merging. Please review the focused PRs above.

Preservation proof remains available at `backup/pr-3403-pre-salvage-20260813` (exact old tip), `recovery/pr-3403-c2-20260813` (fully recovered integration), and the commit-by-commit mapping/checklists in the updated body.
